Milestones/1.6

ollywarren opened the pull request:
**Liveblog 1.6 Milestone**

Addresses the following Issues:

#243 Not Possible to update an entry twice in a session - Thanks to @rebeccahum (PR #297) 
#242 Dangerous Combination of live blog shortcode and key command - Courtesy of Big Bite Creative (This PR)
#241 Remove key events from Key Events Widget does not work after entry update - Thanks to @rebeccahum (PR #297)
#177 Edited Comments appear at the end of live blog after archiving - Thanks to @rebeccahum (PR #297 & #298 )

Adds the following Features:

#87 Auto-archive live blog after certain amount of time - Thanks to @maxk42 (PR #303) Tweaks by Big Bite Creative (This PR)

philipjohn commented:
> #243 Not Possible to update an entry twice in a session - Thanks to @rebeccahum (PR #297)

Verified fix.

> #242 Dangerous Combination of live blog shortcode and key command - Courtesy of Big Bite Creative (This PR)

Verified fix.

> #241 Remove key events from Key Events Widget does not work after entry update - Thanks to @rebeccahum (PR #297)

Verified fix.

> #177 Edited Comments appear at the end of live blog after archiving - Thanks to @rebeccahum (PR #297 & #298 )

Were you able to reproduce the issue on this one? I've just tried and can't reproduce the original issue to verify the fix. I've also asked @rebeccahum

> #87 Auto-archive live blog after certain amount of time - Thanks to @maxk42 (PR #303) Tweaks by Big Bite Creative (This PR)

Could we get the feedback from #303 incorporated for this? Also, I've just tested and it doesn't work for me. Here's what I did;

1. Created a new liveblog
2. Added a few entries
3. Changed the post date to 2017-08-24 and saved
4, Refreshed the liveblog

I expected to see the archive auto-archived, but it was still active.

I verified with `wp shell` that `get_post_time( 'U', true, 12 )` returned `1503583098` (or "2017-08-24 13:58:18").

ollywarren commented:
@philipjohn Ill look into these ASAP, I missed the #303 feedback so will sort this out and see if i can replicate your issues with adjusting the publish date.

With regards #177 I was able to re-create it locally, I used an older branch of the codebase and verified the behaviour, then merged the fix and tested both the command line fix for retrospective bugs and also verified it fixed the issue moving forward. For the sake of clarification Ill make a point of re testing though and verify its working.

ollywarren commented:
@philipjohn to clarify with regards the publish date, the logic doesn't deal with the publish date, so upon creating a new Liveblog post we store the current date + the configured offset (default 7 days) as a timestamp on the options table. Then we check if the current date has passed the "expiry date" and auto archive at that point. By setting the publish date back, you wont effect anything as the expiry date is still set to the original date + offset.

If an archived post, is "re-enabled" as in the user clicks on enable in the admin panel, the same applies. The current date is increased by the offset and that is stored as a timestamp, and when that date is passed the blog is auto-archived again.

Hope that makes sense.

Is there a use case for people moving the publish date of the post back in time? Happy to adjust as directed.

philipjohn commented:
> upon creating a new Liveblog post we store the current date + the configured offset (default 7 days) as a timestamp on the options table.

Ah I'd been looking at the code in #303 so hadn't spotted that. In any case, I think archiving should be dependent on activity. If someone runs a Liveblog for >7 days they're going to be annoyed when it suddenly archives itself mid-flow. In my review of #303 I suggested using the last entry date.

ollywarren commented:
@philipjohn no problems ill update the feature to use that date.

rebeccahum commented:
@philipjohn I was able to reproduce #177.

philipjohn commented on classes/class-wpcom-liveblog-entry.php:
```diff
@@ -281,6 +291,32 @@ private static function user_object_from_comment_id( $comment_id ) {
+		self::$restricted_shortcodes = apply_filters( 'liveblog_entry_restrict_shortcodes', self::$restricted_shortcodes );
+
+		// Foreach lookup key, does it exist in the content.
+		foreach ( self::$restricted_shortcodes as $key => $value ) {
```
We should double check that `self::$restricted_shortcodes` is an array before using it, just in case.

philipjohn commented on classes/class-wpcom-liveblog-entry.php:
```diff
@@ -281,6 +291,32 @@ private static function user_object_from_comment_id( $comment_id ) {
+		foreach ( self::$restricted_shortcodes as $key => $value ) {
+
+			// Regex Pattern will match all shortcode formats.
+			$pattern = "/\[{$key}(.*?)?\](?:(.+?)?\[\/{$key}(.*?)?\])?/";
```
We should replace this with `get_shortcode_regex()` so that we're using whatever core uses to find shortcodes.

ollywarren commented:
@philipjohn Auto Archive now using the latest entry publish date to calculate auto archive date. Each new entry will extend the date by the pre-determined offset. Also removed the options page and implemented a filter to allow developer configuration of the archive days. Docs updated.

Normalized URLs to improve perfromance with caching.

Change for issue #310 - If user/window is not in focus we increase refresh interval to 30 seconds

Change for issue #311 - The refresh interval is passed in the response allowing it to be changed at any time. 

``` 
add_filter( 'liveblog_post_{$post_id}_refresh_interval', function( $refresh_interval ) {
    return 60; 
} ); 
```
```
add_filter( 'liveblog_refresh_interval', function( $refresh_interval ) {
    return 5; 
} );
```

philipjohn commented on classes/class-wpcom-liveblog-wp-cli.php:
```diff
@@ -11,6 +13,105 @@ public function readme_for_github() {
+	 *
+	 * @subcommand fix-archive
+	*/
+	public function fix_archive() {
```
This command is pretty quiet :) We should get some more output so that it's clear to the user what the command is doing. Progress might be useful, too.

philipjohn replied:
Also, a dry-run option would be best.

ollywarren replied:
@philipjohn all the other tweaks are completed Ill tackle this one tomorrow morning first thing .

philipjohn commented on js/liveblog-admin.js:
```diff
@@ -9,17 +9,42 @@ jQuery( function( $ ) {
+			data['template_format']                 = encodeURIComponent( $( '#liveblog-key-template-format' ).val() );
+			data['limit']                           = encodeURIComponent( $( '#liveblog-key-limit' ).val() );
+			data[liveblog_admin_settings.nonce_key] = liveblog_admin_settings.nonce;
+			var method = 'POST';
```
Will this mean all requests from the frontend are `POST` requests? As `POST`s will always be uncached, they'll create unnecessary load if that's the default.

philipjohn replied:
Ignore this :)

ollywarren replied:
👍

philipjohn commented on liveblog.php:
```diff
@@ -332,40 +359,74 @@ public static function ajax_entries_between() {
 		// Do not cache if it's too soon
 		if ( $end_timestamp > time() ) {
-			$response_args['cache'] = false;
+			//$response_args['cache'] = false; --> Potentially Depricated
```
Code that isn't used or needed should be removed completely, rather than commenting it out. If it's needed in the future it can be restored via the Git history.

philipjohn commented on liveblog.php:
```diff
@@ -332,40 +359,74 @@ public static function ajax_entries_between() {
-		foreach( $entries as $entry ) {
-			$latest_timestamp   = max( $latest_timestamp, $entry->get_timestamp() );
-			$entries_for_json[] = $entry->for_json();
+		// if ( empty( $entries ) ) {
```
Code that isn't used or needed should be removed completely, rather than commenting it out. If it's needed in the future it can be restored via the Git history.

philipjohn commented on liveblog.php:
```diff
@@ -332,40 +359,74 @@ public static function ajax_entries_between() {
 
-		self::json_return( $result_for_json, $response_args );
+		return $result;
+		// self::json_return( $result_for_json, $response_args );
```
Code that isn't used or needed should be removed completely, rather than commenting it out. If it's needed in the future it can be restored via the Git history.

philipjohn commented on liveblog.php:
```diff
@@ -395,21 +456,44 @@ public static function is_viewing_liveblog_post() {
+			// expiry is less than today i.e. its in the past lets archive the liveblog.
+			if( $expiry ) {
+				if( (int)$expiry[0] < $today_timestamp ) {
+					self::set_liveblog_state( $post_id, 'archive' );
```
Will this end up running on the front end?

ollywarren replied:
I believe it runs on, all page loads and the original feature PR had it that way. I guess ideally this sort of function would be a CRON job / scheduled task that would run and check once a day or so and run any housekeeping required, is there a preferred way to handle this?

philipjohn replied:
Possibly, for now I'm just concerned with whether this will cause front-end DB writes which would be bad bad bad.

philipjohn commented on liveblog.php:
```diff
@@ -504,6 +588,33 @@ public static function ajax_crud_entry() {
 			), array( 'cache' => false ) );
 		}
+
+		// self::json_return( $entry );
```
Code that isn't used or needed should be removed completely, rather than commenting it out. If it's needed in the future it can be restored via the Git history.

philipjohn commented on liveblog.php:
```diff
@@ -541,30 +672,34 @@ public static function ajax_single_entry() {
+			$result_for_json['previousTimestamp'] = $previous_timestamp;
 
-		self::json_return( $result_for_json );
+			//self::json_return( $result_for_json );
```
Code that isn't used or needed should be removed completely, rather than commenting it out. If it's needed in the future it can be restored via the Git history.

philipjohn commented on liveblog.php:
```diff
@@ -576,52 +711,88 @@ public static function ajax_lazyload_entries() {
-			array( 'html' => $entry_content ),
-			array( 'cache' => false )
-		);
+		// self::json_return(
```
Code that isn't used or needed should be removed completely, rather than commenting it out. If it's needed in the future it can be restored via the Git history.

philipjohn commented on liveblog.php:
```diff
@@ -1275,6 +1609,19 @@ private static function is_wp_too_old() {
+	 *
+	 * @return int
+	 */
+	public static function get_refresh_interval() {
```
Nice!

ollywarren replied:
@philipjohn does this require any action?

ollywarren commented:
@philipjohn just to clarify the settings page, I'm not seeing it in my code base at all? I have removed it for 1.6 can you point me to where your seeing the Settings page?

The metabox on the actual post is still in place, but the settings page that controlled the Auto Archive feature has been removed.

philipjohn commented on classes/class-wpcom-liveblog-wp-cli.php:
```diff
@@ -13,67 +13,96 @@ public function readme_for_github() {
+
+			//Tell the user what we are doing, but lets colour this one se we can see its a new Liveblog in the console output.
+			// Note: Assigned to a varibale to be echo'd out, calling WP_CLI::colorize directly didnt output.
+			$intro = WP_CLI::colorize( "%4 Processing Liveblog {$current_liveblog} of {$total_liveblogs} %n");
```
[WP_CLI::colorize()](https://make.wordpress.org/cli/handbook/internal-api/wp-cli-colorize/) only makes the string fancy, it doesn't output. This needs to be further wrapped by `WP_CLI::line()` rather than echoing.

ollywarren replied:
👍

philipjohn commented on classes/class-wpcom-liveblog-wp-cli.php:
```diff
@@ -13,67 +13,96 @@ public function readme_for_github() {
+							// replace with correct meta_value
+							if ( $correct_ids[$i] < $entry_id ) {
+
+								$wpdb->update(
```
Let's move the dry run check from above to here, so we can report in the dry run on what IDs are changing.

philipjohn commented on classes/class-wpcom-liveblog-wp-cli.php:
```diff
@@ -85,16 +114,16 @@ public function fix_archive() {
-			// check to make sure entry content being replaced matches available
-			if ( count( $entries_replace ) === count( $correct_contents) ) {
+			// check to make sure entry content being replaced matches available and that this is not a dry run
+			if ( count( $entries_replace ) === count( $correct_contents) && false === $is_dryrun ) {
```
Let's move this dry run check until just before the update action, so we can better report on what will be done in live mode.

philipjohn commented:
> just to clarify the settings page, I'm not seeing it in my code base at all? I have removed it for 1.6 can you point me to where your seeing the Settings page?

Ah, it's just the [template](https://github.com/Automattic/liveblog/pull/309/files#diff-1c6c4f85377a6b306bcfbc51ab1189c2R4)

ollywarren commented:
@philipjohn I have now completed the required changes pending your review. Ive stripped out that template file for the settings page as its no longer required and I have moved the Auto Archive expiry check to a CRON job that runs every day at midnight. It was doing a single UPDATE request on the front end so this issues has been addressed and eliminated.

Let me know if theres anything else you need.

philipjohn commented on classes/class-wpcom-liveblog-cron.php:
```diff
@@ -0,0 +1,68 @@
+            $posts = new WP_Query( array(
+                'order'      => 'ASC',
+                'orderby'    => 'ID',
+                'meta_key'   => 'liveblog',
```
Could we use `WPCOM_Liveblog::$auto_archive_expiry_key` here instead?
Also, how many posts will this query return - will it be enough to catch all the Liveblogs we'll need to archive?

ollywarren replied:
@philipjohn ```WPCOM_Liveblog::$auto_archive_expiry_key``` references the key on the meta table that holds the expiry date, in the query above I'm looking for ALL posts that have a liveblog metakey indicating its been activated as a Liveblog post. Then I loop them check any set expiry date and update as required.

If preferred I could just grab only those with a meta_value of 'enabled'? so we skip anything thats already archived.

philipjohn commented on classes/class-wpcom-liveblog-cron.php:
```diff
@@ -0,0 +1,68 @@
+                //if we have an expiry date lets compare them and if the
+                // expiry is less than today i.e. its in the past lets archive the liveblog.
+                if ( $expiry ) {
+                    if ( (int)$expiry[0] < $today ) {
```
Rather than use `$index[0]` could we just set the third parameter on `get_post_meta()` to `true`?

ollywarren replied:
Implemented, makes sense, much simpler.

philipjohn commented on classes/class-wpcom-liveblog-cron.php:
```diff
@@ -0,0 +1,68 @@
+                if ( $expiry ) {
+                    if ( (int)$expiry[0] < $today ) {
+                        WPCOM_Liveblog::set_liveblog_state( $post_id, 'archive' );
+                        $state = get_post_meta( $post_id, WPCOM_Liveblog::key, true );
```
Is this needed?

ollywarren replied:
Ive taken this out.
